srtcore/queue.cpp

```diff
@@ -216,24 +216,28 @@ CUnit *CUnitQueue::getNextAvailUnit()
     if (m_iCount >= m_iSize)
         return NULL;
 
-    CQEntry *entrance = m_pCurrQueue;
-
+    int units_checked = 0;
     do
     {
-        for (CUnit *sentinel = m_pCurrQueue->m_pUnit + m_pCurrQueue->m_iSize - 1; m_pAvailUnit != sentinel;
-             ++m_pAvailUnit)
+        CUnit *lastunit = m_pCurrQueue->m_pUnit + m_pCurrQueue->m_iSize - 1;
+        for (; m_pAvailUnit != lastunit; ++m_pAvailUnit, ++units_checked)
+        {
             if (m_pAvailUnit->m_iFlag == CUnit::FREE)
+            {
                 return m_pAvailUnit;
+            }
+        }
 
-        if (m_pCurrQueue->m_pUnit->m_iFlag == CUnit::FREE)
+        if (lastunit->m_iFlag == CUnit::FREE)
         {
-            m_pAvailUnit = m_pCurrQueue->m_pUnit;
+            m_pAvailUnit = lastunit;
             return m_pAvailUnit;
         }
 
+        ++units_checked;
         m_pCurrQueue = m_pCurrQueue->m_pNext;
         m_pAvailUnit = m_pCurrQueue->m_pUnit;
-    } while (m_pCurrQueue != entrance);
+    } while (units_checked < m_iSize);
```

Comment:
This loop should go through all the units in the queues.
See `TEST(CUnitQueue, IncreaseAndFree)` for the related issue.

Reply:
As I understand the original idea was to break the vicious circle by marking the start point (`entrance`) and check the next unit block pointer against it, while you are using the size as the limit. Does that mean that the circular list not always works, or what was exactly the reason of the change?

Reply:
The `TEST(CUnitQueue, IncreaseAndFree)` provides an example when this condition does not work.
Imagine one queue of 4 units: 0, 1, 2, 3. When unit 3 is taken, units 0, 1 and 2 are free, this condition will not let to find an available unit from the beginning of the same queue.

Reply:
AFAIK this is a circular list, so when you have in effect the current unit 3, and 0 1 2 are free, then the search should start with unit 3 and then roll back to 0, which should find the first free unit there.

Reply:
Consider when unit 3 is found, it is returned via `return m_pAvailUnit;`
This means that the next round of search will start with this very unit in the end of the queue.